### ISSUE TITLE
Feature Request: Parameterized labels for create

### ISSUE THREAD

**Issue by terribleplan** [neo4j/neo4j#4334]
## Description

Allow a user to send a parameter for the label of a create statement.
## Use case

When inserting new nodes into the graph it would allow users to simply `CREATE (n:{type} {document})` to create a new node with a label, instead of requiring them to build a string with the type statically inserted.
## Additional

It'd be great if this could work for non-creation queries as well, since that would still require string building. It may not be possible though, due to constraints in query optimization. I don't know enough to intelligently comment on that end of things, but I would like some additional usefulness (that I actually assumed was already there) out of a tool I am using.

**Comment by cleishm** [comment]
Hi @terribleplan,

Thanks for the input. This is almost a duplicate of #340.

Certainly there are constraints that require knowing the labels in statements that require index selection - use of parameters would hide that from the compiler. Generally, my experience has also been that the desire to parameterize labels/types can be indicative of a modeling concern - a typical model would have a small, consistent and well-known set of labels/types, and if these instead needs to be determined at runtime then it _may_ be that a well-known property and dynamic value is more appropriate.

That said, there are undoubtably some cases that can't be addressed in alternate ways. It would be great if you could share yours.

Cheers,
Chris

**Comment by terribleplan** [comment]
@cleishm Thanks for the quick reply. What I am building is basically a generic middleware for creating a RESTful API in node.js atop Neo4j, and as such I don't know anything about their models until they create one with my middleware. I don't really want to tell users "You can't use 'type' in your model", I would like to give them full access to use whatever properties they want, and then basically manage the rest through whatever metadata-like mechanisms exist.

**Comment by cleishm** [comment]
Have you considered generating the queries to contain the appropriate label then? Middleware for SQL databases also has the same challenge, as you can't parameterize table or column names.

It may be easier in the long run, as you'll want to be substituting in the user-defined labels into read queries also, where index selection is required.

**Comment by terribleplan** [comment]
Yeah, that's what I'm falling back to now, but I have always had an version to string building SQL queries. I figured I would create a feature request in case it could change for the future and I could at least change it in the `CREATE`s.

Feel free to close this in favor of #340, I created this because there seemed to be no comment from contributors on using it in `CREATE`, only a way to use it in `WHERE`.

**Comment by cleishm** [comment]
I'll leave it open for now, and pass the feedback along to our team. Thanks for taking the time to provide it!

**Comment by niroberson** [comment]
I need this

**Comment by nikhil026** [comment]
Has this feature of parameterizing labels in a cypher query  been added ? If not then please guide me a work around for the same . I can't afford to have hard coded Labels.
Further it would be great if someone can explain the idea of not adding this feature even after so long.
@cleishm @terribleplan 



Thanks in Advance 
Cheers
Nikhil

**Comment by cleishm** [comment]
The concern with parametrized labels was that the actual label wouldn’t be available during query planning, and thus the query could not be optimised to use indexes, etc. That could result in serious performance issues for even simple queries, and the risk of that outweighed the benefit. Especially when the need for parametrized labels is more often than not a sign of a problem in the data modeling.

I don’t know if the Neo4j team have come up with a solution since then, as I don’t work with them as often nowadays.

**Comment by boggle** [comment]
Dynamic label parameterization is being worked on as part of Cypher 10 and GQL.

**Comment by lingvisa** [comment]
Hi, Is this feature has been implemented yet?  A common use case for this would be that, you want to create database based on a data model, where data model serves as the parameters for the graph creation code. In that case, labels, relationships and properties will be read from a csv file as usual but they will be all variables, parameters passed to the graph creation function. You don't want to hard code these elements otherwise, you have to re-write your code for each different model, which seems bad. I know that APOC can help create dynamical properties and relationships, but not dynamic labels. This is really needed.

Can the neo4j Team confirm whether is available as of today?

**Comment by sherfert** [comment]
@lingvisa This has not been implemented yet.

**Comment by cellcoresystems** [comment]
@lingvisa I searched like hell and finally found it out.
you can do it like that:

```
// create or update nodes with dynamic label from import data
WITH "file:///query.json" AS url
call apoc.load.json(url) YIELD value as u
UNWIND u.cis as ci
CALL apoc.merge.node([ ci.label ], {Id:ci.Id}, {}, {}) YIELD node
RETURN node;
```

The JSON looks like that:
```
{
   "cis": [
     {
       "label": "Computer",
       "Id": "1"
     },
     {
       "label": "Service",
       "Id": "2"
     },
     {
       "label": "Person",
       "Id": "3"
     }
   ],
   "relations": [
     {
       "end1Id": "1",
       "Id": "4",
       "end2Id": "2",
       "label": "USES"
     },
     {
       "end1Id": "3",
       "Id": "5",
       "end2Id": "1",
       "label": "MANAGED_BY"
     }
   ]
 }
```

**Comment by es50678** [comment]
I need to be able to do something just like what is described in the feature request:
```
CREATE (n:{type} {document})
```

Going to use the workaround @cellcoresystems points out for now.

**Comment by aldearle** [comment]
I would like to be able to do this too, for similar reasons. I am going to generate strings containing the label parameters. This is clearly a hack to get around incompleteness in the design since you could do this for all the parameters not just the labels.

**Comment by process0** [comment]
+1 for this feature.  To note, using `apoc.merge.node/relationship` will not report accurate query updateStatistics

**Comment by mnd999** [comment]
Looks like a duplicate of #2000

**Comment by yachal-upson** [comment]
Still an open issue?? I've ended up in the same situation as other commenters, and it feels like this should have been basic functionality since Neo4j launch. Basically all the data I'll be dealing with will be flat-file with dynamic labels. 

Did apoc functionality just render a fix low priority?

**Comment by jamthief** [comment]
Hi there. Neo4j now fully supports dynamic labels and types as of versions 2025.03 and 5.26.4 LTS, released in March of this year.

You can read more about them [here](https://neo4j.com/blog/developer/cypher-dynamism/), or read the documentation [here](https://neo4j.com/docs/cypher-manual/current/clauses/match/#dynamic-match).

Thanks for your patience and happy MATCHing!
